File: android/src/main/java/io/dolby/sdk/reactnative/mapper/ParticipantMapper.java
package io.dolby.sdk.reactnative.mapper;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;
import com.voxeet.android.media.MediaStream;
import com.voxeet.android.media.stream.AudioTrack;
import com.voxeet.android.media.stream.MediaStreamType;
import com.voxeet.android.media.stream.VideoTrack;
import com.voxeet.sdk.json.ParticipantInfo;
import com.voxeet.sdk.models.Participant;
import com.voxeet.sdk.models.v1.ConferenceParticipantStatus;
import com.voxeet.sdk.models.v2.ParticipantType;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import io.dolby.sdk.reactnative.utils.RNCollectionExtractor;

/**
 * Provides methods that map:
 * <p>- {@link Participant} and {@link Participant}-related models to React Native models</p>
 * <p>- React Native model to {@link Participant} model and {@link Participant}-related models</p>
 */
public class ParticipantMapper {

    public final static String PARTICIPANT_ID = "id";
    public final static String PARTICIPANT_AUDIO_TRANSMITTING = "audioTransmitting";
    public final static String PARTICIPANT_INFO = "info";
    public final static String PARTICIPANT_INFO_NAME = "name";
    public final static String PARTICIPANT_INFO_EXTERNAL_ID = "externalId";
    public final static String PARTICIPANT_INFO_AVATAR_URL = "avatarUrl";
    public final static String PARTICIPANT_STATUS = "status";
    public final static String PARTICIPANT_STREAMS = "streams";
    public final static String PARTICIPANT_STREAMS_ID = "id";
    public final static String PARTICIPANT_STREAMS_TYPE = "type";
    public final static String PARTICIPANT_STREAMS_AUDIO_TRACKS = "audioTracks";
    public final static String PARTICIPANT_STREAMS_VIDEO_TRACKS = "videoTracks";
    public final static String PARTICIPANT_TYPE = "type";

    @NotNull
    private final RNCollectionExtractor rnCollectionExtractor;

    public ParticipantMapper(
            @NotNull RNCollectionExtractor rnCollectionExtractor
    ) {
        this.rnCollectionExtractor = rnCollectionExtractor;
    }

    @NotNull
    public ParticipantInfo toParticipantInfo(@NotNull ReadableMap participantInfoMap) {
        return new ParticipantInfo(
                rnCollectionExtractor.getString(participantInfoMap, PARTICIPANT_INFO_NAME),
                rnCollectionExtractor.getString(participantInfoMap, PARTICIPANT_INFO_EXTERNAL_ID),
                rnCollectionExtractor.getString(participantInfoMap, PARTICIPANT_INFO_AVATAR_URL)
        );
    }

    @Nullable
    public String toParticipantId(@NotNull ReadableMap participantMap) {
        return rnCollectionExtractor.getString(participantMap, PARTICIPANT_ID);
    }

    @NotNull
    public WritableMap toMap(@NotNull Participant participant) {
        WritableMap map = new WritableNativeMap();

        if (participant.getId() != null) {
            map.putString(PARTICIPANT_ID, participant.getId());
        }
        map.putBoolean(PARTICIPANT_AUDIO_TRANSMITTING, participant.audioTransmitting());
        map.putString(PARTICIPANT_STATUS, toString(participant.getStatus()));
        if (participant.getInfo() != null) {
            map.putMap(PARTICIPANT_INFO, toMap(participant.getInfo()));
        }
        map.putArray(PARTICIPANT_STREAMS, toStreamsArray(participant.streams()));
        map.putString(PARTICIPANT_TYPE, toString((participant.participantType())));

        return map;
    }

    @NonNull
    public WritableArray toParticipantsArray(@NotNull List<Participant> participants) {
        WritableNativeArray participantsArray = new WritableNativeArray();
        for (Participant participant : participants) {
            if (participant != null) {
                participantsArray.pushMap(toMap(participant));
            }
        }
        return participantsArray;
    }

    @NotNull
    private String toString(@NotNull ConferenceParticipantStatus participantStatus) {
        switch (participantStatus) {
            case CONNECTING:
                return "CONNECTING";
            case DECLINE:
                return "DECLINE";
            case ERROR:
                return "ERROR";
            case INACTIVE:
                return "INACTIVE";
            case KICKED:
                return "KICKED";
            case LEFT:
                return "LEFT";
            case RESERVED:
                return "RESERVED";
            case WARNING:
                return "WARNING";
            case IN_PROGRESS:
            case ON_AIR:
            case LATER:
            case MISSED:
            case UNKNOWN:
            default:
                return "UNKNOWN";
        }
    }

    @NotNull
    private WritableMap toMap(@NotNull ParticipantInfo participantInfo) {
        WritableMap map = new WritableNativeMap();

        if (participantInfo.getName() != null) {
            map.putString(PARTICIPANT_INFO_NAME, participantInfo.getName());
        }
        if (participantInfo.getExternalId() != null) {
            map.putString(PARTICIPANT_INFO_EXTERNAL_ID, participantInfo.getExternalId());
        }
        if (participantInfo.getAvatarUrl() != null) {
            map.putString(PARTICIPANT_INFO_AVATAR_URL, participantInfo.getAvatarUrl());
        }

        return map;
    }

    @NotNull
    private WritableNativeArray toStreamsArray(@NotNull List<MediaStream> streams) {
        WritableNativeArray streamsArray = new WritableNativeArray();
        for (MediaStream stream : streams) {
            if (stream != null) {
                streamsArray.pushMap(toMap(stream));
            }
        }
        return streamsArray;
    }

    @NotNull
    private ReadableMap toMap(@NotNull MediaStream stream) {
        WritableNativeMap map = new WritableNativeMap();

        map.putString(PARTICIPANT_STREAMS_ID, stream.peerId());
        map.putString(PARTICIPANT_STREAMS_TYPE, toString(stream.getType()));
        map.putArray(PARTICIPANT_STREAMS_AUDIO_TRACKS, toAudioTrackArray(stream.audioTracks()));
        map.putArray(PARTICIPANT_STREAMS_VIDEO_TRACKS, toVideoTrackArray(stream.videoTracks()));

        return map;
    }

    @NotNull
    private String toString(@NotNull MediaStreamType mediaStreamType) {
        switch (mediaStreamType) {
            case Camera:
                return "CAMERA";
            case ScreenShare:
                return "SCREEN_SHARE";
            case Custom1:
            case Custom2:
            case Custom3:
            case Custom4:
            case Custom5:
            case Custom6:
            case Custom7:
            case Custom8:
            default:
                return "UNKNOWN";
        }
    }

    @NotNull
    private ReadableArray toAudioTrackArray(@NotNull List<AudioTrack> audioTracks) {
        WritableNativeArray streamsArray = new WritableNativeArray();
        for (AudioTrack audioTrack : audioTracks) {
            if (audioTrack != null) {
                streamsArray.pushString(audioTrack.id());
            }
        }
        return streamsArray;
    }

    @NotNull
    private ReadableArray toVideoTrackArray(@NotNull List<VideoTrack> videoTracks) {
        WritableNativeArray streamsArray = new WritableNativeArray();
        for (VideoTrack videoTrack : videoTracks) {
            if (videoTrack != null) {
                streamsArray.pushString(videoTrack.id());
            }
        }
        return streamsArray;
    }

    @NotNull
    private String toString(@NotNull ParticipantType participantType) {
        switch (participantType) {
            case USER:
                return "USER";
            case LISTENER:
                return "LISTENER";
            case SPEAKER:
            case PSTN:
            case MIXER:
            case NONE:
            case DVC:
            case ROBOT:
            case ROBOT_SPEAKER:
            case ROBOT_LISTENER:
            case ROBOT_PSTN:
            case ROBOT_MIXER:
            case ROBOT_NONE:
            default:
                return "UNKNOWN";
        }
    }
}

File: android/src/main/java/io/dolby/sdk/reactnative/mapper/ParticipantMapper.kt
package io.dolby.sdk.reactnative.mapper

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableNativeArray
import com.voxeet.android.media.MediaStream
import com.voxeet.android.media.stream.AudioTrack
import com.voxeet.android.media.stream.MediaStreamType
import com.voxeet.android.media.stream.VideoTrack
import com.voxeet.sdk.json.ParticipantInfo
import com.voxeet.sdk.models.Participant
import com.voxeet.sdk.models.v1.ConferenceParticipantStatus
import com.voxeet.sdk.models.v2.ParticipantType

/**
 * Provides methods that map [Participant] and [Participant]-related models to React Native models and vice versa
 */
class ParticipantMapper {

  fun participantIdFromRN(participantRN: ReadableMap): String? =
    participantRN.getString(PARTICIPANT_ID)

  fun infoFromRN(participantInfoRN: ReadableMap) = ParticipantInfo(
    participantInfoRN.getString(PARTICIPANT_INFO_NAME),
    participantInfoRN.getString(PARTICIPANT_INFO_EXTERNAL_ID),
    participantInfoRN.getString(PARTICIPANT_INFO_AVATAR_URL)
  )

  fun toRNInfo(participantInfo: ParticipantInfo): ReadableMap =
    Arguments.createMap().apply {
      participantInfo.name?.let { putString(PARTICIPANT_INFO_NAME, it) }
      participantInfo.externalId?.let { putString(PARTICIPANT_INFO_EXTERNAL_ID, it) }
      participantInfo.avatarUrl?.let { putString(PARTICIPANT_INFO_AVATAR_URL, it) }
    }

  fun toRN(participant: Participant): ReadableMap =
    Arguments.createMap().apply {
      participant.id?.let { putString(PARTICIPANT_ID, participant.id) }
      putBoolean(PARTICIPANT_AUDIO_TRANSMITTING, participant.audioTransmitting())
      putString(PARTICIPANT_STATUS, toRNParticipantStatus(participant.status))
      participant.info?.let { putMap(PARTICIPANT_INFO, toRNInfo(it)) }
      putArray(PARTICIPANT_STREAMS, toRNMediaStreams(participant.streams()))
      putString(PARTICIPANT_TYPE, toRNParticipantType(participant.participantType()))
    }

  fun toRN(participants: List<Participant?>): ReadableArray =
    Arguments.createArray().apply {
      participants
        .filterNotNull()
        .map(::toRN)
        .forEach(::pushMap)
    }

  private fun toRNParticipantType(participantType: ParticipantType) = when (participantType) {
    ParticipantType.USER -> "USER"
    ParticipantType.LISTENER -> "LISTENER"
    ParticipantType.SPEAKER,
    ParticipantType.PSTN,
    ParticipantType.MIXER,
    ParticipantType.NONE,
    ParticipantType.DVC,
    ParticipantType.ROBOT,
    ParticipantType.ROBOT_SPEAKER,
    ParticipantType.ROBOT_LISTENER,
    ParticipantType.ROBOT_PSTN,
    ParticipantType.ROBOT_MIXER,
    ParticipantType.ROBOT_NONE -> "UNKNOWN"
  }

  private fun toRNParticipantStatus(participantStatus: ConferenceParticipantStatus): String =
    when (participantStatus) {
      ConferenceParticipantStatus.CONNECTING -> "CONNECTING"
      ConferenceParticipantStatus.DECLINE -> "DECLINE"
      ConferenceParticipantStatus.ERROR -> "ERROR"
      ConferenceParticipantStatus.INACTIVE -> "INACTIVE"
      ConferenceParticipantStatus.KICKED -> "KICKED"
      ConferenceParticipantStatus.LEFT -> "LEFT"
      ConferenceParticipantStatus.RESERVED -> "RESERVED"
      ConferenceParticipantStatus.WARNING -> "WARNING"
      ConferenceParticipantStatus.IN_PROGRESS,
      ConferenceParticipantStatus.ON_AIR,
      ConferenceParticipantStatus.LATER,
      ConferenceParticipantStatus.MISSED,
      ConferenceParticipantStatus.UNKNOWN -> "UNKNOWN"
    }

  private fun toRNMedialStream(stream: MediaStream): ReadableMap =
    Arguments.createMap().apply {
      putString(PARTICIPANT_STREAMS_ID, stream.peerId())
      putString(PARTICIPANT_STREAMS_TYPE, toRNMediaStreamType(stream.type))
      putArray(PARTICIPANT_STREAMS_AUDIO_TRACKS, toRNAudioTracks(stream.audioTracks()))
      putArray(PARTICIPANT_STREAMS_VIDEO_TRACKS, toRNVideoTracks(stream.videoTracks()))
    }

  private fun toRNMediaStreams(streams: List<MediaStream>): ReadableArray =
    Arguments.createArray().apply {
      streams
        .map(::toRNMedialStream)
        .forEach(::pushMap)
    }

  private fun toRNMediaStreamType(mediaStreamType: MediaStreamType): String =
    when (mediaStreamType) {
      MediaStreamType.Camera -> "CAMERA"
      MediaStreamType.ScreenShare -> "SCREEN_SHARE"
      else -> "UNKNOWN"
    }

  private fun toRNAudioTracks(audioTracks: List<AudioTrack?>): ReadableArray =
    WritableNativeArray().apply {
      audioTracks
        .filterNotNull()
        .map(AudioTrack::id)
        .forEach(::pushString)
    }

  private fun toRNVideoTracks(videoTracks: List<VideoTrack?>): ReadableArray =
    Arguments.createArray().apply {
      videoTracks
        .filterNotNull()
        .map(VideoTrack::id)
        .forEach(::pushString)
    }

  companion object {
    private const val PARTICIPANT_ID = "id"
    private const val PARTICIPANT_AUDIO_TRANSMITTING = "audioTransmitting"
    private const val PARTICIPANT_INFO = "info"
    private const val PARTICIPANT_INFO_NAME = "name"
    private const val PARTICIPANT_INFO_EXTERNAL_ID = "externalId"
    private const val PARTICIPANT_INFO_AVATAR_URL = "avatarUrl"
    private const val PARTICIPANT_STATUS = "status"
    private const val PARTICIPANT_STREAMS = "streams"
    private const val PARTICIPANT_STREAMS_ID = "id"
    private const val PARTICIPANT_STREAMS_TYPE = "type"
    private const val PARTICIPANT_STREAMS_AUDIO_TRACKS = "audioTracks"
    private const val PARTICIPANT_STREAMS_VIDEO_TRACKS = "videoTracks"
    private const val PARTICIPANT_TYPE = "type"
  }
}
